### PR TITLE
z-index | Tools | Map, function, and mixin refactor

### PR DESCRIPTION
**Pull request by mikemai2awesome**
## Jira

http://vjira2:8080/browse/BDS-659

## Summary

Refactored z-index settings and tools in `bolt-core`.

## Details

1. New map is created using auto index method
2. Function and mixin usage stay the same, deprecation warnings are in place for all deprecated z-index names
3. All existing z-index usage has been converted to use the new z-index names
4. Doc page is revamped to clearly document how to use the util classes and the function
5. A decision tree graphic has been made to help understand the complexity of z-index

## How to test

Run the branch locally and go through the documentation page (http://localhost:3000/pattern-lab/?p=visual-styles-z-index). Create a dummy page and some dummy CSS to test out the `bolt-z-index()` function, make sure the function is working as expected according to the stacking order specified in the map.

### CODE REVIEW

**Comment by christophersmith262** [comment]
@mikemai2awesome just wanted to document this so you are aware.  Independent of our theme, Drupal core has its own z-index hierarchy to consider.

Here's all the places I've found that could end up getting used on a Drupal page:
```
./misc/dropbutton/dropbutton.css:  z-index: 100;
./misc/dialog/off-canvas.dropbutton.css:  z-index: 1;
./misc/dialog/off-canvas.dropbutton.css:  z-index: 100;
./misc/dialog/off-canvas.theme.css:  z-index: 501;
./misc/dialog/off-canvas.reset.css:  z-index: 1;
./misc/dialog/off-canvas.button.css:  z-index: 10;
./themes/stable/css/core/dropbutton/dropbutton.css:  z-index: 100;
./themes/stable/css/core/dialog/off-canvas.dropbutton.css:  z-index: 1;
./themes/stable/css/core/dialog/off-canvas.dropbutton.css:  z-index: 100;
./themes/stable/css/core/dialog/off-canvas.theme.css:  z-index: 501;
./themes/stable/css/core/dialog/off-canvas.reset.css:  z-index: 1;
./themes/stable/css/core/dialog/off-canvas.button.css:  z-index: 10;
./themes/stable/css/contextual/contextual.theme.css:  z-index: 500;
./themes/stable/css/contextual/contextual.theme.css:  z-index: 2;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 98;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 99;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 300;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 300;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 100;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: 100;
./themes/stable/css/quickedit/quickedit.module.css:  z-index: -1;
./themes/stable/css/quickedit/quickedit.theme.css:  z-index: 2;
./themes/stable/css/quickedit/quickedit.theme.css:  z-index: 1;
./themes/stable/css/quickedit/quickedit.theme.css:  z-index: 0;
./themes/stable/css/views_ui/views_ui.admin.css:  z-index: 51;
./themes/stable/css/views_ui/views_ui.admin.css:  z-index: 50;
./themes/stable/css/views_ui/views_ui.admin.theme.css:  z-index: 2;
./themes/stable/css/toolbar/toolbar.menu.css:  z-index: 1;
./themes/stable/css/toolbar/toolbar.module.css:  z-index: 1250;
./themes/stable/css/toolbar/toolbar.module.css:    z-index: -999;
./themes/stable/css/toolbar/toolbar.module.css:  z-index: 502;
./themes/stable/css/toolbar/toolbar.module.css:  z-index: 501;
./themes/stable/css/toolbar/toolbar.module.css:  z-index: -1;
./themes/stable/css/toolbar/toolbar.module.css:  z-index: 502;
./themes/stable/css/ckeditor/ckeditor.admin.css:  z-index: 11; /* Ensure focused buttons show their outline on all sides. */
./themes/stable/css/ckeditor/ckeditor.admin.css:  z-index: 10;
./themes/stable/css/ckeditor/ckeditor.admin.css:  z-index: 10;
./themes/stable/css/system/components/ajax-progress.module.css:  z-index: 1000;
./themes/stable/css/system/components/sticky-header.module.css:  z-index: 500;
./themes/stable/css/tour/tour.module.css:  z-index: 101;
./themes/stable/css/tour/tour.module.css:  z-index: 100;
./themes/stable/css/tour/tour.module.css:  z-index: 102;
./themes/stable/css/tour/tour.module.css:  z-index: 10000;
./themes/stable/css/node/node.preview.css:  z-index: 499;
./themes/stable/css/block/block.admin.css:  z-index: 499;
./themes/stable/css/simpletest/simpletest.module.css:  z-index: 1000;
```

File paths are relative to https://git.drupalcode.org/project/drupal/tree/8.8.x/core

At least a few of those z-index's are pretty important, like the toolbar and generic 'ui-dialog' class.

**Comment by margoromo** [comment]
Thanks Chris for the input!

**Comment by mikemai2awesome** [comment]
@christophersmith262 

Thank you! It's great to see all that.

Regardless of what system we have for Bolt, there will be potential conflicts with those drupal interface z-indexes (or any other CSS code base Bolt would interact with, if neither are encapsulated). The new system in this PR will result in [low numbers](https://boltdesignsystem-bepawdxyvb.now.sh/pattern-lab/?p=visual-styles-z-index), so chances for conflicts are lower than the current system.

This new system will make sure at least the components we produce will be consistent in terms of stacking order. And if custom work is created using the Bolt function or utility classes (and the developer follows the guidelines we provide), this system will keep them consistent as well.

File: docs-site/src/pages/pattern-lab/_patterns/01-visual-styles/20-utility-classes/z-index.twig [hunk 1]
```diff
@@ -1,32 +1,128 @@
...
+      first_col_fixed_width: true,
+    } only %}
+
+    {% include "@bolt-components-headline/headline.twig" with {
+      text: "Usage Guidelines",
+      size: "xxlarge",
+      tag: "h2",
+    } only %}
+    {% include "@bolt-components-headline/text.twig" with {
+      text: "The usage of z-index is very complicated, do not use it blindly and without context. Use <code>content-low</code>, <code>content</code>, and <code>content-low</code> to handle general stacking order within a container. Use <code>nav</code>, <code>overlay</code>, <code>floating-low</code>, <code>floating</code>, <code>floating-high</code>, <code>modal-overlay</code>, <code>modal</code> for more specific high level UI components that are fixed or absolute position.",
```
Comment (adamszalapski) [inline docs-site/src/pages/pattern-lab/_patterns/01-visual-styles/20-utility-classes/z-index.twig]: This fragment `Use <code>content-low</code>, <code>content</code>, and <code>content-low</code>` should be changed to `Use <code>content-low</code>, <code>content</code>, and <code>content-high</code>`

**Comment by sghoweri** [comment]
@mikemai2awesome @danielamorse I can take another look at this once @danielamorse's PR feedback is addressed + merge conflicts resolved

**Comment by mikemai2awesome** [comment]
@sghoweri @danielamorse @adamszalapski 

Your comments are addressed now.

**Comment by colbytcook** [comment]
Work being tracked in https://pegadigitalit.atlassian.net/browse/DS-110

**Pull request closed without merging**